### PR TITLE
Optimize connections

### PR DESCRIPTION
**Pull request by mundya**
Connections are now added to a `ConnectionMap` object when they are built. This structure ensures that equivalent connections (e.g., those that share sufficient signal parameters and transforms/decoders/etc.) are simulated on SpiNNaker by only one stream of multicast packets.

I'd recommend any review starts from `nengo_spinnaker/builder/model.py`.

### CODE REVIEW

**Comment by mundya** [comment]
Any feedback would be appreciated... Can I ask for thoughts please @tcstewar / @neworderofjamie.

Note this PR fixes #49.

**Comment by mundya** [comment]
@neworderofjamie, are you still happy to do a joint code-review of this early next week?

**Comment by neworderofjamie** [comment]
Yes I am - Today or tomorrow are probably optimal!

File: nengo_spinnaker/builder/builder.py [hunk 1]
```diff
@@ -45,10 +45,8 @@ class Model(object):
         Map of objects to the operators which will simulate them on SpiNNaker.
     extra_operators: [operator, ...]
         Additional operators.
-    connections_signals : {connection: :py:`~.Signal`, ...}
-        Map of connections to the signals that simulate them.
-    extra_signals: [operator, ...]
-        Additional signals.
+    connection_map :
+        Connection manager.
```
Comment (neworderofjamie) [inline nengo_spinnaker/builder/builder.py]: Has this been renamed / maybe describe it's role

File: nengo_spinnaker/regions/filters.py [hunk 2]
```diff
@@ -35,21 +35,22 @@ def make_filter_regions(signals_and_connections, dt, minimise=False,
     filters = list()
     keyspace_routes = list()
 
-    for signal, connections in iteritems(signals_and_connections):
-        for connection in connections:
-            # Make the filter
-            f = FilterRegion.supported_filter_types[type(connection.synapse)].\
-                from_signal_and_connection(signal, connection, width=width)
+    for signal, rps in specs:
```
Comment (neworderofjamie) [inline nengo_spinnaker/regions/filters.py]: Being pedantic, I think `rps` is overly terse  personally I'd call it's `reception_params`, but even `params` would make it's role clearer

File: nengo_spinnaker/operators/value_source.py [hunk 3]
```diff
@@ -212,10 +209,10 @@ def write_subregion_to_file(self, fp, vertex_slice, **kwargs):
         ))
 
 
-def get_transform_keys(model, sig, conn):
+def get_transform_keys(sig, tps):
```
Comment (neworderofjamie) [inline nengo_spinnaker/operators/value_source.py]: Again pedantry, but I have no idea what a `tps` is

Comment (mossblaser) [inline nengo_spinnaker/operators/value_source.py]: http://www.urbandictionary.com/define.php?term=tps+report&defid=518450

File: nengo_spinnaker/operators/value_source.py [hunk 4]
```diff
@@ -39,20 +39,17 @@ def make_vertices(self, model, n_steps):
 
         # Get all the outgoing signals to determine how big the size out is and
         # to build a list of keys.
-        sigs_conns = model.get_signals_connections_from_object(self)
+        sigs_conns = model.get_signals_from(self)
```
Comment (neworderofjamie) [inline nengo_spinnaker/operators/value_source.py]: this is what I meant by the objective c like renaming :)

File: nengo_spinnaker/operators/sdp_receiver.py [hunk 5]
```diff
@@ -27,22 +28,18 @@ def make_vertices(self, model, *args, **kwargs):
 
         # Get all outgoing signals and their associated connections (this
         # SHOULD be a 1:1 mapping)
-        out = model.get_signals_connections_from_object(self)
-        for signal, connections in six.iteritems(out[OutputPort.standard]):
-            assert len(connections) == 1, "Expecting a 1:1 mapping"
-            conn = connections[0]
-
+        for signal, tps in model.get_signals_from(self)[OutputPort.standard]:
```
Comment (neworderofjamie) [inline nengo_spinnaker/operators/sdp_receiver.py]: I still hate `tps` (counter-intuitively, I'm reviewing backwards) especially as, according to the comment, it's a connection....

Comment (mundya) [inline nengo_spinnaker/operators/sdp_receiver.py]: You want me to keep comments up to date with the surrounding code?

Madness.

File: nengo_spinnaker/builder/model.py [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+                (self.keyspace == other.keyspace)):
+            return True
+        else:
+            return False
+
+    def __ne__(self, b):
+        return not self == b
+
+
+class TransmissionParameters(object):
```
Comment (neworderofjamie) [inline nengo_spinnaker/builder/model.py]: None of the classes I (by their names) should derive from this actually seem to! Should this class just be deleted or is magic occurring somewhere I missed?

**Pull request merged**